src/main/java/dk/cngroup/trainings/spring/springassignment/service/AnimalServiceImpl.java

```diff
@@ -1,80 +1,109 @@
-        }
-    }
+	@Override
+	public Optional<Animal> addAnimal(Animal animal) {
+		if (AnimalValidationService.isValid(animal)) {
+			animal.setId(IdGenerator.getId());
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
+		}
+	}
 
-    @Override
-    public boolean deleteAnimalById(long id) {
-        if(animalRepository.existsById(id)) {
-            animalRepository.deleteById(id);
-            return true;
-        }else
+	@Override
+	public boolean deleteAnimalById(long id) {
+		if (animalRepository.existsById(id)) {
+			animalRepository.deleteById(id);
+			return true;
+		} else
 
-            return false;
-    }
+		{
+			return false;
+		}
+	}
 
-    @Override
-    public Optional<Animal> updateAnimalById(long id, @Valid Animal animal) { //TODO delete @Valid
-        if(isValid(animal) && animalRepository.existsById(id)) {
-            // Change the updated animal id and override it in the database.
-            animal.setId(id);
-            return Optional.ofNullable(animalRepository.save(animal));
-        } else {
-            return null; //TODO Optional.empty();
-        }
-    }
+	@Override
+	public Optional<Animal> updateAnimalById(long id, Animal animal) {
+		if (AnimalValidationService.isValid(animal) && animalRepository.existsById(id)) {
+			// Change the updated animal id and override it in the database.
+			animal.setId(id);
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
+		}
+	}
 
-    @Override
-    public boolean isValid(@Valid Animal animal) { //it would be possible to move this e.g. to AnimalValidationService
-        //TODO check also nullability
-        return animal.getName().length()>=Animal.NAME_MINIMUM_SIZE
-                && animal.getDescription().length()<Animal.DESCRIPTION_MAXIMUM_SIZE
-                && !animal.getDescription().toLowerCase().contains("penguin");
-    }
+	@Override
+	public Optional<CareTaker> addNewCareTakerToExistingAnimal(long id, CareTaker careTaker) {
+		Optional<Animal> animal = animalRepository.findById(id);
+		if (CareTakerValidationService.isValid(careTaker)) {
+			careTaker.setId(IdGenerator.getId());
+			List<Animal> animals = Arrays.asList(animal.get());
+			careTaker.setAnimals(animals);
+			CareTaker addedCareTaker = careTakerRepository.save(careTaker);
+			return Optional.ofNullable(addedCareTaker);
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	@Override
+	public String addExistingCareTakerToExistingAnimal(long animalId, long careTakerId) {
+		Optional<Animal> animal = animalRepository.findById(animalId);
+		Optional<CareTaker> careTaker = careTakerRepository.findById(careTakerId);
+		if (!animal.isPresent()) {
+			return "Failed: Animal id doesn't exist";
+		} else if (!careTaker.isPresent()) {
+			return "Failed: CareTaker id doesn't exist";
+		} else {
+			if (animal.get().getCareTakers().stream().anyMatch(c -> c.getId() == careTakerId)) {
```

Comment:
if careTakers were saved in a Set, then you would not need to worry about the duplicates, right? just keep in mind you need to have equals and hashcode methods correctly implemented

Reply:
Yeah, Sets don't allow duplicates.
I tried migrating the collection from List to Set but it turned out to be a lot of refactoring. Too much time consuming that would make keeping that check for duplicates and testing more convenient.